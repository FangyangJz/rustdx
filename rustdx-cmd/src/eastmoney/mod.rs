use eyre::{Result, WrapErr};
use serde::{Deserialize, Serialize};

/// 获取股票数据
pub fn get(n: u16) -> Result<String> {
    // 如果需要升序，使用 `order=code%2Case` 或者 `order=`
    // ashare => A 股，bshare => B 股，kshare => 科创板，equity => 前三种
    let url = format!(
        "http://56.push2.eastmoney.com/api/qt/clist/get?\
            cb=jQuery112407375845698232317_1631693257414&\
            pn=1&pz={n}&po=0&np=1&ut=bd1d9ddb04089700cf9c27f6f7426281&\
            fltt=2&invt=2&fid=f12&fs=m:0+t:6,m:0+t:80,m:1+t:2,m:1+t:23&\
            fields=f18,f16,f12,f17,f15,f2,f6,f5&_=1631693257534",
    );
    Ok(ureq::get(&url)
        .call()
        .wrap_err_with(|| format!("获取东财股票数据失败，网址为\n`{url:?}`"))?
        .into_string()?)
}

pub fn parse(text: &str) -> Result<EastMarket> {
    // jQuery112407375845698232317_1631693257414();
    serde_json::from_str(&text[42..text.len() - 2])
        .wrap_err_with(|| format!("解析东财股票数据失败，返回的文本为\n{text:?}"))
}

/// 用于（反）序列化：比如读取东方财富网页返回的 json ；把结果写入到 csv
/// 注意：factor 需要提供前一天的 factor 数据才会计算（即 -p xx.csv）
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Day<'a> {
    /// `date` 为 `%Y-%m-%d` 文本格式
    #[serde(skip_deserializing, default = "default_date")]
    pub date: String,
    #[serde(rename(deserialize = "f12"))]
    pub code: String,
    #[serde(borrow)]
    #[serde(rename(deserialize = "f17"))]
    pub open: F32<'a>,
    #[serde(borrow)]
    #[serde(rename(deserialize = "f15"))]
    pub high: F32<'a>,
    #[serde(borrow)]
    #[serde(rename(deserialize = "f16"))]
    pub low: F32<'a>,
    #[serde(borrow)]
    #[serde(rename(deserialize = "f2"))]
    pub close: F32<'a>,
    #[serde(borrow)]
    #[serde(rename(deserialize = "f6"))]
    pub amount: F32<'a>,
    #[serde(borrow)]
    #[serde(rename(deserialize = "f5"))]
    pub vol: F32<'a>,
    #[serde(borrow)]
    #[serde(rename(deserialize = "f18"))]
    pub preclose: F32<'a>,
    #[serde(skip_deserializing, default)]
    pub factor: f64,
}

/// 排除掉 "-" 无实际数据的股票
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum F32<'a> {
    Null(&'a str),
    Yes(f32),
}

/// TODO： 最新的交易日，而不是当天
fn default_date() -> String {
    chrono::Local::now().format("%Y-%m-%d").to_string()
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EastMarket<'a> {
    #[serde(borrow)]
    pub data: EastData<'a>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EastData<'a> {
    // pub diff:  Vec<Factor>,
    #[serde(borrow)]
    pub diff: Vec<Day<'a>>,
    pub total: u16,
}
