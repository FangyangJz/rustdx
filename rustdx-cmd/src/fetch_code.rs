use eyre::{anyhow, Result};
use std::collections::HashSet;

// 股票上限
const SH8: &str = "800";
const SH1: &str = "2000";

pub type StockList = HashSet<String>;

#[derive(Debug)]
pub struct SHSZ {
    sh1: usize,
    sh8: usize,
    sz: usize,
}

impl SHSZ {
    /// 计算总和
    pub fn count(&self) -> usize {
        let SHSZ { sh1, sh8, sz } = &self;
        sh1 + sh8 + sz
    }
}

/// 获取上证和深证的股票代码
///
/// 数量如下：SHSZ { sh1: 1646, sh8: 407, sz: 2745, } -> 4798
///
/// 见 `tests-integration::fetch_code::offical_stocks` 测试
pub fn offical_stocks(set: &mut StockList) -> Result<SHSZ> {
    let count = SHSZ {
        sh1: get_sh_stocks(set, "1", SH1)?,
        sh8: get_sh_stocks(set, "8", SH8)?,
        sz: get_sz_stocks(set)?,
    };
    info!("股票数量 {count:?}");
    Ok(count)
}

pub fn get_offical_stocks(cond: &str) -> Result<StockList> {
    let mut set = StockList::with_capacity(6000);
    let len = match cond {
        "official" => offical_stocks(&mut set)?.count(),
        "szse" => get_sz_stocks(&mut set)?,
        "sse" => get_sh_stocks(&mut set, "8", SH8)? + get_sh_stocks(&mut set, "1", SH1)?,
        _ => unreachable!("请输入 official | szse | sse 中的一个"),
    };

    info!("获得上证和深证股票数量：{len}");
    Ok(set)
}

/// 深交所官网的 A 股和创业板股票信息。
pub fn get_sz_stocks(set: &mut StockList) -> Result<usize> {
    use calamine::{DataType, Reader, Xlsx};
    use std::io::Read;
    let (url, ex) = (
        "http://www.szse.cn/api/report/ShowReport?\
        SHOWTYPE=xlsx&CATALOGID=1110&TABKEY=tab1&random=0.8587844061443386",
        "sz",
    );
    let bytes = &mut Vec::with_capacity(1 << 20);
    ureq::get(url).call()?.into_reader().read_to_end(bytes)?;
    let mut workbook = Xlsx::new(std::io::Cursor::new(bytes))?;
    // 每个单元格被解析的类型可能会不一样，所以把股票代码统一转化成字符型
    if let Some(Ok(range)) = workbook.worksheet_range_at(0) {
        set.extend(range.rows().skip(1).map(|r| match &r[4] {
            DataType::Int(x) => format!("{ex}{x}"),
            DataType::Float(x) => {
                format!("{}{}", ex, *x as i64)
            }
            DataType::String(x) => format!("{ex}{x}"),
            _ => unreachable!(),
        }));
        Ok(range.height() - 1)
    } else {
        Err(anyhow!("xlsx parse error"))
    }
}

const USER_AGENT: &str = "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.107 Safari/537.36";
const ACCEPT_LANGUAGE: &str =
    "zh-CN,zh;q=0.9,de;q=0.8,ko;q=0.7,ru;q=0.6,it;q=0.5,ga;q=0.4,en;q=0.3";

pub fn ureq_sz_with_headers(url: &str) -> ureq::Request {
    ureq::get(url)
        .set("ACCEPT", "*/*")
        .set("ACCEPT_LANGUAGE", ACCEPT_LANGUAGE)
        .set("CACHE_CONTROL", "no-cache")
        .set("CONNECTION", "keep-alive")
        .set("CONTENT_TYPE", "application/json")
        .set("DNT", "1")
        .set("PRAGMA", "no-cache")
        .set("REFERER", "http://www.sse.com.cn/")
        .set("USER_AGENT", USER_AGENT)
}
pub fn ureq_sh_with_headers(url: &str) -> ureq::Request {
    let cookie = "ba17301551dcbaf9_gdp_user_key=; \
                  ba17301551dcbaf9_gdp_session_id=0876b773-38a3-44d0-bb4a-2b5569025b82; \
                  gdp_user_id=gioenc-2g8894g6%2C764a%2C50d8%2C8d6g%2C3bg6194752ce; \
                  ba17301551dcbaf9_gdp_session_id_0876b773-38a3-44d0-bb4a-2b5569025b82=true; \
                  JSESSIONID=0311A5533F5FD798EE9DAFDE6A1D70A7; \
                  ba17301551dcbaf9_gdp_sequence_ids=\
                  {%22globalKey%22:14%2C%22VISIT%22:2%2C%22PAGE%22:5%2C%22VIEW_CHANGE%22:2%2C%22CUSTOM%22:3%2C%22VIEW_CLICK%22:6}";
    ureq::get(url)
        .set("ACCEPT", "*/*")
        .set("ACCEPT_LANGUAGE", ACCEPT_LANGUAGE)
        .set("CACHE_CONTROL", "no-cache")
        .set("CONNECTION", "keep-alive")
        .set("COOKIE", cookie)
        .set("PRAGMA", "no-cache")
        .set("REFERER", "http://www.sse.com.cn/")
        .set("USER_AGENT", USER_AGENT)
}

// 上交所 科创板 68 开头（目前 350 只，只需一次请求） => stockType=8, pagesize=400
//        A 股 60 开头（目前 1650 只，只需一次请求） => stockType=1, pagesize=1700
fn request_sh(stocktype: &str, pagesize: &str) -> ureq::Request {
    let url = format!(
        "http://query.sse.com.cn/sseQuery/commonQuery.do?\
        jsonCallBack=jsonpCallback37525685&STOCK_TYPE={stocktype}\
        &REG_PROVINCE=&CSRC_CODE=&STOCK_CODE=&sqlId=COMMON_SSE_CP_GPJCTPZ_GPLB_GP_L\
        &COMPANY_STATUS=2%2C4%2C5%2C7%2C8&type=inParams&isPagination=true\
        &pageHelp.cacheSize=1&pageHelp.beginPage=1&pageHelp.pageSize={pagesize}\
        &pageHelp.pageNo=1&pageHelp.endPage=1&_=1680491539414"
    );
    ureq_sh_with_headers(&url)
}

// 上交所 科创板 68 开头（目前 350 只，只需一次请求） => stockType=8, pagesize=400
//        A 股 60 开头（目前 1650 只，只需一次请求） => stockType=1, pagesize=1700
pub fn get_sh_stocks(set: &mut StockList, stocktype: &str, pagesize: &str) -> Result<usize> {
    let text = request_sh(stocktype, pagesize).call()?.into_string()?;
    let pos1 = text
        .find("total\":")
        .ok_or(anyhow!("`Total` field not found"))?
        + 7;
    let pos2 = text[pos1..pos1 + 10]
        .find('}')
        .ok_or(anyhow!("`Total` field not found"))?
        + pos1;
    let n: usize = text[pos1..pos2].parse()?;
    // 注意：如果不 take 的话，split 有一半是重复的

    set.extend(
        text.split("COMPANY_CODE")
            .skip(1)
            .take(n)
            .map(|s| format!("sh{}", &s[3..9])),
    );
    Ok(n)
}
