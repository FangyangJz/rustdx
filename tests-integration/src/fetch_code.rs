#[test]
fn offical_stocks() {
    use crate::{now, shot, snap};
    use rustdx_cmd::fetch_code::{offical_stocks, StockList};

    shot!(now(), @"2023-02-23 15:50:16.053645411 +08:00");

    let mut set = StockList::with_capacity(6000);
    let (count, time) = elapse!(offical_stocks(&mut set).unwrap());
    let len = count.count();
    assert_eq!(set.len(), len);

    snap!(count, @r###"
    SHSZ {
        sh1: 1646,
        sh8: 407,
        sz: 2745,
    }
    "###);
    shot!(len, @"4798");
    shot!(time, @"795");
}
