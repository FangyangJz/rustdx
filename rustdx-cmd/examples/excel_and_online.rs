#![feature(once_cell)]
use rustdx_cmd::fetch_code;
use std::sync::LazyLock;

macro_rules! get {
    (sz) => {{
        let mut set = ::std::collections::HashSet::with_capacity(3000);
        fetch_code::get_sz_stocks(&mut set).unwrap();
        let mut v = set.into_iter().collect::<Vec<_>>();
        v.sort();
        v
    }};
    (sh, $a:literal, $b:literal) => {{
        let mut set = ::std::collections::HashSet::with_capacity(3000);
        fetch_code::get_sh_stocks(&mut set, $a, $b).unwrap();
        let mut v = set.into_iter().collect::<Vec<_>>();
        v.sort();
        v
    }};
}

static SH8: LazyLock<Vec<String>> = LazyLock::new(|| get!(sh, "8", "1000"));
static SH1: LazyLock<Vec<String>> = LazyLock::new(|| get!(sh, "1", "2500"));
static SZ: LazyLock<Vec<String>> = LazyLock::new(|| get!(sz));

/// sh8: 334
/// ["sh688001", "sh688002", "sh688003", "sh688004", "sh688005", "sh688006", "sh688007",
///  "sh688008", "sh688009", "sh688010"]
/// ["sh688787", "sh688788", "sh688789", "sh688793", "sh688798", "sh688799", "sh688800",
///  "sh688819", "sh688981", "sh689009"]
/// sh1: 1639
/// ["sh600000", "sh600004", "sh600006", "sh600007", "sh600008", "sh600009", "sh600010",
///  "sh600011", "sh600012", "sh600015"]
/// ["sh605398", "sh605399", "sh605488", "sh605499", "sh605500", "sh605507", "sh605577",
///  "sh605580", "sh605588", "sh605589"]
/// sz: 2488
/// ["sz000001", "sz000002", "sz000004", "sz000005", "sz000006", "sz000007", "sz000008",
///  "sz000009", "sz000010", "sz000011"]
/// ["sz301045", "sz301046", "sz301047", "sz301048", "sz301049", "sz301050", "sz301051",
///  "sz301052", "sz301053", "sz301055"]
fn main() {
    let (sh8, sh1, sz) = (&*SH8, &*SH1, &*SZ);
    println!(
        "sh8: {}\n{:?}\n{:?}",
        sh8.len(),
        &sh8[..10],
        &sh8[sh8.len() - 10..]
    );
    println!(
        "sh1: {}\n{:?}\n{:?}",
        sh1.len(),
        &sh1[..10],
        &sh1[sh1.len() - 10..]
    );
    println!(
        "sz: {}\n{:?}\n{:?}",
        sz.len(),
        &sz[..10],
        &sz[sz.len() - 10..]
    );
}

#[test]
fn save() -> eyre::Result<()> {
    use insta::assert_debug_snapshot as snap;
    use rustdx_cmd::eastmoney;
    use std::collections::HashSet;

    let (sh8, sh1, sz) = (&*SH8, &*SH1, &*SZ);
    let (lsh8, lsh1, lsz) = (sh8.len(), sh1.len(), sz.len());
    snap!("sh1", sh1);
    snap!(lsh1, @"1679");
    snap!("sh8", sh8);
    snap!(lsh8, @"510");
    snap!("sz", sz);
    snap!(lsz, @"2758");
    let l = lsh1 + lsh8 + lsz;
    snap!(l, @"4947");

    let s = eastmoney::get(6000)?;
    let res = eastmoney::parse(&s)?;
    let east: HashSet<_> = res
        .data
        .diff
        .into_iter()
        .filter(|v| matches!(v.open, eastmoney::F32::Yes(_)))
        .map(|v| v.code)
        .collect();
    let mut v = east.iter().collect::<Vec<_>>();
    v.sort();
    let lv = v.len();
    let total = res.data.total as usize;
    eyre::ensure!(lv == total, "lv = {lv} 与 total = {total} 不相等");
    snap!("eastmoney", v);
    snap!(lv, @"5168");
    snap!(lv == l, @"false");

    let exchange = HashSet::from_iter(
        [sh8.iter().cloned(), sh1.iter().cloned(), sz.iter().cloned()]
            .into_iter()
            .flatten()
            .map(|s| s[2..].to_string()),
    );
    snap!("diff_exchange-east", &exchange - &east);
    snap!("diff_east-exchange", &east - &exchange);

    Ok(())
}
