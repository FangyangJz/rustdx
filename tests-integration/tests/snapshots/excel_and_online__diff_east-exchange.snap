---
source: rustdx-cmd/examples/excel_and_online.rs
expression: "&east - &exchange"
---
{}
